### DISCUSSION TITLE
I'm using a SX1276 board.  spi initializes and radio

### DISCUSSION THREAD

**Discussion by DanielW0830** [jgromes/RadioLib#1365] [General]
```
 myspi->begin(MY_SPI_SCK,MY_SPI_MISO,MY_SPI_MOSI,MY_SPI_NSS);

      Serial.print(F("[SX1276] Initializing ... "));
      int state = radio.begin();
      if (state == RADIOLIB_ERR_NONE) 
      {
        g_bUsingLora = true;
        Serial.println(F("success!"));
      } 
```

this code returns success.

later in the code I do
 `transmissionState = radio.startTransmit("Sample:Text");`

and in loop I do
  ```
 // check if the previous transmission finished
  if(transmittedFlag) 
  {
    // reset flag
    transmittedFlag = false;

    if (transmissionState == RADIOLIB_ERR_NONE) 
    {
      // packet was successfully sent
      Serial.println(F("transmission finished!"));

      // NOTE: when using interrupt-driven transmit method,
      //       it is not possible to automatically measure
      //       transmission data rate using getDataRate()

    } else 
    {
      Serial.print(F("failed, code "));                                        **<=====  I get this with error    -16**
      Serial.println(transmissionState);

    }
```
Here's my pins:
const int MY_SPI_MOSI = 33;
const int MY_SPI_MISO = 23;
const int MY_SPI_SCK = 14;
const int MY_SPI_NSS = 5;
const int MY_LORA_RESET = 2;
const int MY_LORA_DIO0 = 26;

Any ideas?

**Comment by jgromes** [comment]
I would start with [debug mode](https://github.com/jgromes/RadioLib/wiki/Debug-mode) with `RADIOLIB_DEBUG_SPI` enabled, that will tell you exactly which register is failing. Also the complete code will be useful - it should be the minimum code that still manifests the issue, but complete.

**Reply by DanielW0830** [reply]
I uncommented the debug spi line, did a clean and no debug showed up.   What did I do wrong?   Thanks.

**Reply by DanielW0830** [reply]
strange that spi initialized and the radiolib initialized and then I get SPI error on send.
System is powered by a power supply (not batteries)
some messages on-line say to put a large cap on power supply to SX1276.   Is this true?
Maybe the voltage is dipping during the send.

**Reply by jgromes** [reply]
> I uncommented the debug spi line

The one in BuildOptUser.h right? And you did rebuild and upload this new version, correct?

> strange that spi initialized and the radiolib initialized and then I get SPI error on send

RadioLib performs a lot of checks to ensure everything is going as it should, this includes checking all writes to SPI registers. This error means that after writing to a register, the value that was read out of it does not match what was written. The debug log will include more information, such as which exact register this was and the affected bits.

> Maybe the voltage is dipping during the send

That is also my suspicion, but I want to confirm this with the debug log. If this really is the cause, we will see failure to start the Tx mode. Also you did not mention which microcontroller you are using, that would be helpful to know as well.

**Reply by DanielW0830** [reply]
I'm using a ESP32 WROOM 32E

**Reply by DanielW0830** [reply]
my regulator is a RT9080-33GJ5  (max current is 600ma)
dropping 4.5v to 3.3v
Regulator docs say:
The RT9080 features an independent current limiter that monitors and controls the gate voltage of the pass
transistor, limiting the output current to 1.1A (typical). When the output voltage decreases further, the current
limiting level is reduced to around 0.6A, which is named fold-back current limit. This allows the output to be shorted
to ground indefinitely without damaging the device.

I could put a 220uf on the output pin of that.

**Reply by DanielW0830** [reply]
added the 220uf on output of  3.3v regulator.   still get -16 on send.

**Comment by DanielW0830** [comment]
here's the debug from the failed send.

RLB_SPI: R      1       89
RLB_SPI: R      1       89
RLB_SPI: R      1       89
RLB_SPI: R      24      0
RLB_SPI: R      40      0
RLB_SPI: W      40      40
RLB_SPI: R      40      40
RLB_SPI: R      1       89
RLB_SPI: W      12      FF
RLB_SPI: R      22      1
RLB_SPI: W      22      16
RLB_SPI: R      22      16
RLB_SPI: R      E       80
RLB_SPI: W      E       0
RLB_SPI: R      E       0
RLB_SPI: R      D       0
RLB_SPI: W      0       4D      61      69      6C      50      61      74      72      6F      6C      3A      64 6F       6F      72      3A      6F      70      65      6E      65      64
RLB_SPI: R      1       89
RLB_SPI: R      1       89
RLB_SPI: W      1       8B
RLB_SPI: R      1       F
RLB_SPI: R      1       F
RLB_SPI: R      1       F
RLB_SPI: R      1       F
RLB_SPI: R      1       F
RLB_SPI: R      1       F
RLB_SPI: R      1       F
RLB_SPI: R      1       F
RLB_SPI: R      1       F
RLB_SPI: R      1       F
RLB_SPI:
RLB_SPI: address:       0x1
RLB_SPI: bits:          2 0
RLB_SPI: value:         0x3
RLB_SPI: current:       0x89
RLB_SPI: mask:          0x7
RLB_SPI: new:           0x8B
RLB_SPI: read:          0xF

RLB_DBG: -16 at lib/RadioLib/src/modules/SX127x/SX127x.cpp:640
Send Complete.
failed, code -16

**Reply by DanielW0830** [reply]
```
int16_t SX127x::setMode(uint8_t mode) {
  uint8_t checkMask = 0x07;
  if((getActiveModem() == RADIOLIB_SX127X_FSK_OOK) && (mode == RADIOLIB_SX127X_RX)) {
    // disable checking of RX bit in FSK RX mode, as it sometimes seem to fail (#276)
    checkMask = 0xFE;
  }
  return(this->mod->SPIsetRegValue(RADIOLIB_SX127X_REG_OP_MODE, mode, 2, 0, 100, checkMask));
}
```
I noticed the checkmask for mode in SX127x.cpp was an 0xFF.   Shouldn't it be a 0x07 ?
I also increased the timeout on the checkinterval to 100ms (yes very long, but still didnt work)
I changed that and get:

RLB_SPI: address:       0x1     <== mode
RLB_SPI: bits:          2 0         <== bits to modify  
RLB_SPI: value:         0x3      <=== value we want in there
RLB_SPI: current:       0x89   <== current value in mode register
RLB_SPI: mask:          0x7     <== mask applied
RLB_SPI: new:           0x8B    <== new value to write (8B would signify lora mode & low freq + TX mode )  (this is correct)
RLB_SPI: read:          0x9       <== new value read and verified.  0x09 means it switched to FSK/OOK mode and is in standby.
RLB_DBG: -16 at lib/RadioLib/src/modules/SX127x/SX127x.cpp:640
failed, code -16

very weird.

I did run this test with a 220uf cap on the power to the lora chip.

**Reply by jgromes** [reply]
> I noticed the checkmask for mode in SX127x.cpp was an 0xFF. Shouldn't it be a 0x07 ?

No, why? By default we want to check all the bits in the register, not just the changed ones.

Both of the SPI debug logs to me would suggest the device is resetting when going to Tx, because both times, the bit that signals LoRa mode (which is set very early during `begin()`) is reset. That seems to suggest a hardware issue.

220 uF is a huge capacitance -  the datasheet for RT9080-33GJ5 lists 1 uF in the lectrical characteristics table.
